File: go/problems/subrectangle_queries.go
package problems

// https://leetcode.com/problems/subrectangle-queries
type SubrectangleQueries struct {
	Rectangle [][]int
}

func NewSubrectangle(rectangle [][]int) SubrectangleQueries {
	return SubrectangleQueries{Rectangle: rectangle}
}

func (sq *SubrectangleQueries) UpdateSubrectangle(row1 int, col1 int, row2 int, col2 int, newValue int) {
	for i := row1; i <= row2; i++ {
		for j := col1; j <= col2; j++ {
			sq.Rectangle[i][j] = newValue
		}
	}
}

func (sq *SubrectangleQueries) GetValue(row int, col int) int {
	return sq.Rectangle[row][col]
}

File: rust/src/problems/subrectangle_queries.rs
// https://leetcode.com/problems/subrectangle-queries
pub trait Query {
    fn new(rectangle: Vec<Vec<i32>>) -> Self;
    fn update_subrectangle(&mut self, row1: i32, col1: i32, row2: i32, col2: i32, new_value: i32);
    fn get_value(&self, row: i32, col: i32) -> i32;
}

#[derive(Debug)]
pub struct SubrectangleQueries {
    rectangle: Vec<Vec<i32>>
}

impl Query for SubrectangleQueries {
    fn new(rectangle: Vec<Vec<i32>>) -> Self {
        return SubrectangleQueries {
            rectangle
        };
    }

    fn update_subrectangle(&mut self, row1: i32, col1: i32, row2: i32, col2: i32, new_value: i32) {
        for i in row1..=row2 {
            for j in col1..=col2 {
                self.rectangle[i as usize][j as usize] = new_value
            }
        }
    }

    fn get_value(&self, row: i32, col: i32) -> i32 {
        return self.rectangle[row as usize][col as usize];
    }
}
